File: fisheye_window.py
import numpy as np
import cv2
import sys

usage_text = '''
usage:
  python fisheye360.py your-fisheye-image-file
'''
howto_text = '''
Hit followings to move your view:
  'r' or 'f' to zoom
  'g' or 't' to rotate alpha
  'h' or 'y' to rotate beta
  'j' or 'u' to rotate theta
  's' to save current view to ./result.png

Hit ESC to exit.
'''

class FishEyeWindow(object):
    """Fisheye Window Class
    Get 
    
    Args:
        nx, ny (int): Number of inner corners of the chessboard pattern, in x and y axes.
    """
    def __init__(
            self,
            srcWidth,
            srcHeight,
            destWidth,
            destHeight
        ):
        # Initial parameters
        self._srcW = srcWidth
        self._srcH = srcHeight
        self._destW = destWidth
        self._destH = destHeight
        self._al = 0
        self._be = 0
        self._th = 0
        self._R  = srcWidth / 2
        self._zoom = 1.0
        # Map storage
        self._mapX = np.zeros((self._destH, self._destW), np.float32)
        self._mapY = np.zeros((self._destH, self._destW), np.float32)
    def buildMap(self, alpha=None, beta=None, theta=None, R=None, zoom=None):
        # Set the angle parameters
        self._al = (alpha, self._al)[alpha == None]
        self._be = (beta, self._be)[beta == None]
        self._th = (theta, self._th)[theta == None]
        self._R = (R, self._R)[R == None]
        self._zoom = (zoom, self._zoom)[zoom == None]
        # Build the fisheye mapping
        al = self._al / 180.0
        be = self._be / 180.0
        th = self._th / 180.0
        A = np.cos(th) * np.cos(al) - np.sin(th) * np.sin(al) * np.cos(be)
        B = np.sin(th) * np.cos(al) + np.cos(th) * np.sin(al) * np.cos(be)
        C = np.cos(th) * np.sin(al) + np.sin(th) * np.cos(al) * np.cos(be)
        D = np.sin(th) * np.sin(al) - np.cos(th) * np.cos(al) * np.cos(be)
        mR = self._zoom * self._R
        mR2 = mR * mR
        mRsinBesinAl = mR * np.sin(be) * np.sin(al)
        mRsinBecosAl = mR * np.sin(be) * np.cos(al)
        centerV = int(self._destH / 2.0)
        centerU = int(self._destW / 2.0)
        centerY = int(self._srcH / 2.0)
        centerX = int(self._srcW / 2.0)
        # Fill in the map
        for absV in range(0, int(self._destH)):
            v = absV - centerV
            vv = v * v
            for absU in range(0, int(self._destW)):
                u = absU - centerU
                uu = u * u
                upperX = self._R * (u * A - v * B + mRsinBesinAl)
                lowerX = np.sqrt(uu + vv + mR2)
                upperY = self._R * (u * C - v * D - mRsinBecosAl)
                lowerY = lowerX
                x = upperX / lowerX + centerX
                y = upperY / lowerY + centerY
                _v = (v + centerV, v)[centerV <= v]
                _u = (u + centerU, u)[centerU <= u]
                self._mapX.itemset((_v, _u), x)
                self._mapY.itemset((_v, _u), y)

    def getImage(self, img):
        # Look through the window
        output = cv2.remap(img, self._mapX, self._mapY, cv2.INTER_LINEAR)
        return output

def main(input_file):
    src_img = cv2.imread(input_file)
    src_size = src_img.shape[:2]
    cv2.imshow('Original', src_img)
    print('source image size = ' + str(src_size))

    fe = FishEyeWindow(src_size[1], src_size[0], 320, 320)
    alpha = -270
    beta = 0
    theta = 270
    zoom = 1.0
    while True:
        fe.buildMap(alpha=alpha, beta=beta, theta=theta, zoom=zoom)
        result_img = fe.getImage(src_img)
        cv2.imshow('Result', result_img)
        key = cv2.waitKey(0)

        if key == 27:
            break
        elif ord('r') == key:
            zoom -= 0.1
        elif ord('f') == key:
            zoom += 0.1
        elif ord('g') == key:
            alpha += 90
        elif ord('t') == key:
            alpha -= 90
        elif ord('h') == key:
            beta += 45
        elif ord('y') == key:
            beta -= 45
        elif ord('j') == key:
            theta += 90
        elif ord('u') == key:
            theta -= 90
        elif ord('s') == key:
            cv2.imwrite('./result.png', result_img)
        print('alpha=%f, beta=%f, theta=%f, zoom=%f' % (alpha, beta, theta, zoom))
        
    cv2.destroyAllWindows()
    

if __name__ == '__main__':
    if len(sys.argv) <= 1:
        print usage_text
    else:
        print howto_text
        main(sys.argv[1])

File: cpp/fisheye_window.cpp
#include "opencv2/core.hpp"
#include "opencv2/imgproc.hpp"
#include "opencv2/highgui.hpp"
#include <stdio.h>
#include <vector>
#include <array>
#include <cmath>
#include <cassert>

using namespace std;

#define NUMOF_STOREABLE_MAPS 3

class FishEyeWindow {
private:
	int srcW_, srcH_, destW_, destH_;
	float al_, be_, th_, R_, zoom_;
	vector<cv::Mat *> mapXs_, mapYs_;
public:
	
    FishEyeWindow(int srcW, int srcH, int destW, int destH)
	: srcW_(srcW), srcH_(srcH), destW_(destW), destH_(destH),
	  al_(0), be_(0), th_(0), R_(srcW / 2.0), zoom_(1.0),
	  mapXs_(NUMOF_STOREABLE_MAPS, NULL), mapYs_(NUMOF_STOREABLE_MAPS, NULL) {}

	~FishEyeWindow() {
		array<vector<cv::Mat *> *, 2> maps = {&mapXs_, &mapYs_};
		for (int i = 0; i < maps.size(); i++) {
			for (vector<cv::Mat *>::iterator it = maps[i]->begin(); it != maps[i]->end(); it++) {
				delete *it;
			}
		}
	}

	void buildMap(float alpha, float beta, float theta, float zoom, int idx = 0) {
		assert(0 <= idx && idx < NUMOF_STOREABLE_MAPS);
		cv::Mat *mapX = new cv::Mat(destH_, destW_, CV_32FC1);
		cv::Mat *mapY = new cv::Mat(destH_, destW_, CV_32FC1);
		// # Set the angle parameters
		al_ = alpha;
		be_ = beta;
		th_ = theta;
		//R_ = R(R, R_)[R == None]
		zoom_ = zoom;
		// # Build the fisheye mapping
		float al = al_ / 180.0;
		float be = be_ / 180.0;
        float th = th_ / 180.0;
        float A = cos(th) * cos(al) - sin(th) * sin(al) * cos(be);
		float B = sin(th) * cos(al) + cos(th) * sin(al) * cos(be);
        float C = cos(th) * sin(al) + sin(th) * cos(al) * cos(be);
        float D = sin(th) * sin(al) - cos(th) * cos(al) * cos(be);
		float mR = zoom_ * R_;
		float mR2 = mR * mR;
		float mRsinBesinAl = mR * sin(be) * sin(al);
		float mRsinBecosAl = mR * sin(be) * cos(al);
		int centerV = int(destH_ / 2.0);
		int centerU = int(destW_ / 2.0);
		float centerY = srcH_ / 2.0;
		float centerX = srcW_ / 2.0;
		// # Fill in the map
		for (int absV = 0; absV < destH_; absV++) {
			float v = absV - centerV;
			float vv = v * v;
			for (int absU = 0; absU < destW_; absU++) {
				float u = absU - centerU;
				float uu = u * u;
				float upperX = R_ * (u * A - v * B + mRsinBesinAl);
				float lowerX = sqrt(uu + vv + mR2);
				float upperY = R_ * (u * C - v * D - mRsinBecosAl);
				float lowerY = lowerX;
				float x = upperX / lowerX + centerX;
				float y = upperY / lowerY + centerY;
				int _v = centerV <= v ? v : v + centerV;
				int _u = centerU <= u ? u : u + centerU;
				mapX->at<float>(_v, _u) = x;
				mapY->at<float>(_v, _u) = y;
			}
		}
		// # Append as new map
		if (mapXs_[idx] != NULL) delete mapXs_[idx];
		if (mapYs_[idx] != NULL) delete mapYs_[idx];
		mapXs_[idx] = mapX;
		mapYs_[idx] = mapY;
	}

	void getImage(cv::Mat &src, cv::Mat &dest, int idx) {
		cv::remap(src, dest, *mapXs_[idx], *mapYs_[idx],
				  CV_INTER_LINEAR, cv::BORDER_CONSTANT, cv::Scalar(0, 0, 0));
	}
};

static void errorExit(const char *format, const char *arg)
{
    fprintf(stderr, "Error: ");
    fprintf(stderr, format, arg);
    fprintf(stderr, "\nexit.\n");
    exit(-1);
}

int main(int ac, char *av[])
{
    if (ac <= 1) {
		printf("Usage: ./%s your-fisheye-image-file\n", av[0]);
		return -1;
    }

	cv::Mat src_img = cv::imread(av[1]);
	if (!src_img.data) {
		errorExit("No such file: %s", av[1]);
		return 0;
    }
	cv::namedWindow("Original");
	cv::namedWindow("Result");
	cv::imshow("Original", src_img);
	cv::Mat dest_img(320, 320, src_img.type());

    FishEyeWindow few(src_img.rows, src_img.cols, 320, 320);
    float alpha = -270.0;
    float beta = 0.0;
    float theta = 270.0;
    float zoom = 1.0;

	printf("Hit followings to move your view:\n"
		   "'r' or 'f' to zoom\n"
		   "'g' or 't' to rotate alpha\n"
		   "'h' or 'y' to rotate beta\n"
		   "'j' or 'u' to rotate theta\n"
		   "'s' to save current view to ./result.png\n"
		   "\n"
		   "Hit ESC to exit.\n");

	while (true) {
		few.buildMap(alpha, beta, theta, zoom);
		few.getImage(src_img, dest_img, 0);
		cv::imshow("Result", dest_img);

		int key = cv::waitKey(0);
        if (key == 27) {
			break;
		}
		switch (key) {
			case 'r':	zoom -= 0.1; break;
			case 'f':	zoom += 0.1; break;
			case 'g':	alpha += 90; break;
			case 't':	alpha -= 90; break;
			case 'h':	beta += 45; break;
			case 'y':	beta -= 45; break;
			case 'j':	theta += 90; break;
			case 'u':	theta -= 90; break;
			case 's':	cv::imwrite("./result.png", dest_img); break;
		}
        printf("alpha=%f, beta=%f, theta=%f, zoom=%f\n", alpha, beta, theta, zoom);
	}

	cv::destroyAllWindows();

	return 0;
}
